r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+        // C.y = lambda(A.x - C.x) - A.y
+        //
+        // Special cases:
+        //
+        // doubling: if B.y = A.y and B.x = A.x then lambda is unbound and
+        // C = (lambda^2, lambda^3)
+        //
+        // addition of negative point: if B.y = -A.y and B.x = A.x then no
+        // lambda can satisfy the first equation unless B.y - A.y = 0. But
+        // then this reduces to doubling.
+        //
+        // So we need to check that A.x - B.x != 0, which can be done by
+        // enforcing I * (B.x - A.x) = 1
+        if other.is_zero() {
+            return Err(SynthesisError::AssignmentMissing);
+        }
+        let other = other.into_affine();
+        let other_x = other.x;
+        let other_y = other.y;
+
+        let x2_minus_x1 = self
+            .x
+            .sub_constant(cs.ns(|| "x2 - x1"), &other_x)?
+            .negate(cs.ns(|| "neg1"))?;
+        let y2_minus_y1 = self
+            .y
+            .sub_constant(cs.ns(|| "y2 - y1"), &other_y)?
+            .negate(cs.ns(|| "neg2"))?;
+
+        let inv = x2_minus_x1.inverse(cs.ns(|| "compute inv"))?;
+
+        let lambda = NonNativeFieldGadget::alloc(cs.ns(|| "lambda"), || {
+            Ok(y2_minus_y1.get_value().get()? * &inv.get_value().get()?)
+        })?;
+
+        let x_3 = NonNativeFieldGadget::alloc(&mut cs.ns(|| "x_3"), || {
+            let lambda_val = lambda.get_value().get()?;
+            let x1 = self.x.get_value().get()?;
+            let x2 = other_x;
+            Ok((lambda_val.square() - &x1) - &x2)
+        })?;
+
+        let y_3 = NonNativeFieldGadget::alloc(&mut cs.ns(|| "y_3"), || {
+            let lambda_val = lambda.get_value().get()?;
+            let x_1 = self.x.get_value().get()?;
+            let y_1 = self.y.get_value().get()?;
+            let x_3 = x_3.get_value().get()?;
+            Ok(lambda_val * &(x_1 - &x_3) - &y_1)
+        })?;
+
+        // Check lambda
+        lambda.mul_equals(cs.ns(|| "check lambda"), &x2_minus_x1, &y2_minus_y1)?;
+
+        // Check x3
+        let x3_plus_x1_plus_x2 = x_3
+            .add(cs.ns(|| "x3 + x1"), &self.x)?
+            .add_constant(cs.ns(|| "x3 + x1 + x2"), &other_x)?;
+        lambda.mul_equals(cs.ns(|| "check x3"), &lambda, &x3_plus_x1_plus_x2)?;
+
+        // Check y3
+        let y3_plus_y1 = y_3.add(cs.ns(|| "y3 + y1"), &self.y)?;
+        let x1_minus_x3 = self.x.sub(cs.ns(|| "x1 - x3"), &x_3)?;
+
+        lambda.mul_equals(cs.ns(|| ""), &x1_minus_x3, &y3_plus_y1)?;
+
+        Ok(Self::new(x_3, y_3, Boolean::Constant(false)))
+    }
+
+    /// Variable base exponentiation.
+    /// Inputs must be specified in *little-endian* form.
+    /// If the addition law is incomplete for the identity element,
+    /// `result` must not be the identity element.
+    fn mul_bits<'a, CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        result: &Self,
+        bits: impl Iterator<Item = &'a Boolean>,
+    ) -> Result<Self, SynthesisError> {
```

Comment:
Let's add docu how the circuit treats exceptions due to incomplete arithemtics.
Also here: why not using the shift method ?

Reply:
We still need to implement it for the native stuff. We will update this once we do it, and we will use the shift method directly in the application code